### ISSUE TITLE
Unable to read token during refresh using the PHP AWS SDK

### ISSUE THREAD

**Issue by DZDomi** [aws/amazon-eks-pod-identity-webhook#34]
**What happened**:

We are getting the following error when using the php aws sdk:

`Warning: file_get_contents(/var/run/secrets/eks.amazonaws.com/serviceaccount/token): failed to open stream: No such file or directory`

This only happens once a day, exactly when the jwt token is expired and gets refreshed. I checked it with the `exp` field.

**What you expected to happen**:

Being able to still read the file during the refresh of the token

**How to reproduce it (as minimally and precisely as possible)**:

Deploy identity webook into your cluster, add the correct eks annotations to the service account and make queries with the aws sdk to any aws service during the refresh period of the token

**Anything else we need to know?**:

--- 

**Environment**:
- AWS Region: eu-west-1
- EKS Platform version (if using EKS, run `aws eks describe-cluster --name <name> --query cluster.platformVersion`): eks.6
- Kubernetes version (if using EKS, run `aws eks describe-cluster --name <name> --query cluster.version`): 1.14
- Webhook Version: git hash: 2fd4becd434f92edab01cecee30ea4b12af93ab7

**Comment by micahhausler** [comment]
Hmm I think this sounds like a Kubernetes bug where the token isn't being refreshed. 

* Are you able to reliably reproduce this? 
* How long have you encountered this?
* Is that the full error message? 
* Can you post the pod volume spec that the webhook added?
* Can you post the output of  `kubectl get --raw /version`?
* For the node this is occurring on, can you post the output of  
  ```
  kubectl get node $NODE  -o jsonpath='{.status.nodeInfo.kubeletVersion}'
  ```

<details><summary> links for myself to remember</summary>

* [AWS PHP SDK AssumeRoleWithWebIdentityCredentialProvider](https://github.com/aws/aws-sdk-php/blob/47d667bc9b853fc44307a1217c432af99c4fa1f8/src/Credentials/AssumeRoleWithWebIdentityCredentialProvider.php#L14)
* [K8s Projected token volume source](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/projected/projected.go)
</details>

**Comment by DZDomi** [comment]
- Yes this can be reliably reproduced. The pod gets a few requests per second and if it its during that time period of the token refresh the container is unable to access the file for the token, which results in an Access denied exception for the AwsClient we are using in the SDK.
- This only happens in the our prod environment and we encountered it since we have it in use in our production system (4 days). There obv the load to our servers is different then in staging, which makes it not showing up at exactly the refresh times of the token.
- Well its a warning, which then results in a the corresponding AwsClient (for example SQS) throwing an exception.
- Could you show me a command to get the pod yaml?
- Output:
```
{
  "major": "1",
  "minor": "14+",
  "gitVersion": "v1.14.9-eks-c0eccc",
  "gitCommit": "c0eccca51d7500bb03b2f163dd8d534ffeb2f7a2",
  "gitTreeState": "clean",
  "buildDate": "2019-12-22T23:14:11Z",
  "goVersion": "go1.12.12",
  "compiler": "gc",
  "platform": "linux/amd64"
}
```
- Output:
```
v1.14.7-eks-1861c5
```

Thanks for the fast response

**Comment by micahhausler** [comment]
> Yes this can be reliably reproduced. The pod gets a few requests per second and if it its during that time period of the token refresh the container is unable to access the file for the token, which results in an Access denied exception for the AwsClient we are using in the SDK.

Do you have an exportable reproduction? Like a simple PHP application in a container (not your actual application) that you can re-trigger the behavior with?

> Well its a warning, which then results in a the corresponding AwsClient (for example SQS) throwing an exception.

Can you post the warning and the exception? I also want to check that the PHP SDK is failing in the correct way here

> Could you show me a command to get the pod yaml?

```
kubectl get pod -o yaml --export $POD
```

**Comment by micahhausler** [comment]
I've tried to build a minimal reproduction, but was unable to. Here's my horrible stab at PHP. Feel free to correct anything or make it more similar to your setup

https://gist.github.com/micahhausler/7b276124cd4e70ae28dca36a36756cbf

**Comment by DZDomi** [comment]
Hey sorry i do not have time for this yet, i will look into it the next days. Thanks for the info so far!

**Comment by mf-lit** [comment]
We've been suffering this same issue for several weeks now. Just like @DZDomi it's occurring for us in production but not non-prod. Although I simply put this down to the fact that a much higher volume of AWS API calls in production means it's more likely to make a call at the time the token is refreshing.

The issue happens across all nodes, but it doesn't look like it necessarily happens on every renewal, although it's hard to prove that is definitely the case as I can't be sure our application is making API calls at token refresh time.

Exception:
```
/srv/vendor/aws/aws-sdk-php/src/Credentials/InstanceProfileProvider.php:240
Error retrieving credentials from the instance profile metadata server. (cURL error 7: Failed to connect to 169.254.169.254 port 80: Connection refused (see https://curl.haxx.se/libcurl/c/libcurl-errors.html))
```

Warning:
```
/srv/vendor/aws/aws-sdk-php/src/Credential/AssumeRoleWithWebIdentityCredentialProvider.php:97
Warning: file_get_contents(/var/run/secrets/eks.amazonaws.com/serviceaccount/token): failed to open stream: No such file or directory
```

I originally tried to reproduce it by using a very simple Daemonset that ran a shell script that checked the existence of the token file every second. I ran that for a month and didn't see the issue occur a single time.

I then found this issue and ran a [slightly modified](https://gist.github.com/mf-lit/bedba723a3907efd7dd6338b89a5557b) version of your code as a standalone pod @micahhausler 

I ran that for three days (with token renewal set at 10 minutes) and again couldn't reproduce the issue at all.

My EKS platform is 8 and K8S version is 1.14. Tomorrow I'm intending to upgrade to 1.15 in the vague hope that might help.

@DZDomi I wonder did you ever resolve the issue?
@micahhausler If there is anything else you think might be worth checking please do let me know.


As a slight aside, one other thing I noticed is that the token file refresh time is surprisingly uneven. I've graphed some of them below. We can see that all the renewals take place after 480 seconds (as expected, 600*80%=480) but often they take a lot longer than that. Unlikely to be relevant here, unless perhaps sometimes the renewal doesn't happen before the token expires.

![image](https://user-images.githubusercontent.com/9348839/80397819-a577cb80-88ae-11ea-8255-f3bf10c4a597.png)

**Comment by DZDomi** [comment]
hey @mf-lit 

Thanks for helping out with diagnosing. 

We are still encountering this issue, also after upgrade to EKS 1.15. Currently we did not have enough time to investigate the issue as we can also only reproduce it on our production system and its currently non critical. But it still happens about every day, independent of the service. Most of the services which are affected by this get a lot of request/per second, which i think is similar to your workload i guess from @mf-lit 

I will try to put some time into this or probably next week

Maybe its also important to notice that we are running a self deployed version of the pod identity webhook as at that time our clusters were already created and did not have the webhook automatic from AWS included

**Comment by jqmichael** [comment]
@DZDomi and @mf-lit 

Could you send me your cluster ARN and a timestamp when this happened? We need to go dig into our logs.

**Comment by mf-lit** [comment]
Hi @jqmichael , thanks for getting in touch.

I've got the information ready to send you, but I'm hesitating as posting ARNs in a public place like this feels insecure (though I've no idea what attack vector this might provide). Is there a way I can message you directly?

**Comment by mf-lit** [comment]
We believe we've root-caused this now. PHP caches files by default, resulting in it using a stale token.

Some good background can be found in this article https://pracucci.com/php-realpath-cache-and-kubernetes-secrets-configmap-updates.html

And here is an open issue with suggested fix for the AWS SDK: https://github.com/aws/aws-sdk-php/issues/2014

So I'm almost certain this issue can be closed.

**Comment by jqmichael** [comment]
> Hi @jqmichael , thanks for getting in touch.
> 
> I've got the information ready to send you, but I'm hesitating as posting ARNs in a public place like this feels insecure (though I've no idea what attack vector this might provide). Is there a way I can message you directly?

Emm, Github doesn't support private message yet. 
https://github.community/t5/How-to-use-Git-and-GitHub/How-to-sned-a-private-message-to-another-Git/m-p/33856

Next time you can PM in EKS slack channel.

**Comment by jqmichael** [comment]
> We believe we've root-caused this now. PHP caches files by default, resulting in it using a stale token.
> 
> Some good background can be found in this article https://pracucci.com/php-realpath-cache-and-kubernetes-secrets-configmap-updates.html
> 
> And here is an open issue with suggested fix for the AWS SDK: [aws/aws-sdk-php#2014](https://github.com/aws/aws-sdk-php/issues/2014)
> 
> So I'm almost certain this issue can be closed.

Nice finding!
